Finish full text search on a course

ptn opened the pull request:
Hi Jordan,

I started where nerian left off and have fixed a bug. It happens when you pass an empty array as the second argument to texticle's #search, which results in an unscoped search. For example, if you search for

Course.search("test", some_course.assignments)

but some_course does not have any assignments, the method would search in all of the assignments in the database.

My ideas for next are:
- Provide a global search box located next to the link to the public site, so that users can search from anywhere.
- Show in the results screen the text snippet where the match was found. This will probably involve creating a TextSearchResult model.
- Highlight the keyword in said snippet.

How does that sound?

Nerian committed:
Implementing Course text-search. Search controller returns an array with the findings.

Nerian committed:
Implement text-search in all resources of a course

Nerian committed:
Refactor

Nerian committed:
Load the results with Javascript

Nerian committed:
Perform Ajax request

Nerian committed:
highlight description and shared notes

Nerian committed:
Fix the condition

Nerian committed:
Test Course#search_course_resources

Nerian committed:
simplify the factory

ptn committed:
Fix unscoped texticle searches

The #search method that texticle adds takes as a second param an array
of elements that are the only ones to be searched. However, if this is
empty, instead of not performing a search, texticle will ignore it and
search in an unscoped manner.

ptn committed:
Minor wording clarification

jordanbyron commented:
Hey @ptn thanks for continuing @Nerian's work on this feature.

Here are some answers:

> Provide a global search box located next to the link to the public site, so that users can search from anywhere.

Let's just focus on providing search within the context of a course. So I think that means you'll want to add a search box to `app/views/courses/show` which searches within the current course. IIRC `Course#search_course_resources` does that.

> Show in the results screen the text snippet where the match was found. This will probably involve creating a TextSearchResult model.

Can you give me an overview of what the `TextSearchResult` class would do for us?

> Highlight the keyword in said snippet.

I believe we have a helper which does this already. Take a look at the chat channel search code.

---

Looks like you've still got a lot of work to do. Let me know if you need any help :email:

ptn commented:
> Can you give me an overview of what the TextSearchResult class would do for us?

My idea is to show in #search-results a snippet of text for all types of results (course description, comments, assignments, etc.), to provide context. Currently, the full course description is shown with the search keyword highlighted, but this should be a snippet and every type of search result should provide it.

We are using a hash that maps result types to an array to store matching objects classified by type of result, like so:

```
results = {}
results[:type1] = Type1.search(keyword, scope1)
results[:type2] = Type2.search(keyword, scope2)
```

We would need one more level of nesting to insert the snippet in this hash, like so:

```
results = {}

type1 = []

Type1.search(keyword, scope1).each do |match|
  result = {}
  result[:object] = match
  result[:snippet] = get_snippet(match)

  type1 << result
end

results[:type1] = type1

# same for Type2

```

So the new data structure would be a hash of arrays of hashes. Things start to get messy. The TextSearchResult class would wrap a search results and take care of extracting the snippet given the keyword, like this:

```
results = TextSearchResults.new("keyword")
results.add Type1.search("keyword")
results.add Type2.search("keyword")
results[:type1].each do |match|
  # You can access the object itself
  match.object
  # Or the snippet, which TextSearchResults takes care of extracting
  match.snippet
end
```

How about it?

ptn commented:
> Let's just focus on providing search within the context of a course. So I think that means you'll want to add a search box to app/views/courses/show which searches within the current course. IIRC Course#search_course_resources does that.

This is what I meant, I think my choice of words was not clear. 

I wouldn't like to require users to go to the course/show view and then use the sidebar to finally start searching. The idea of having the search form next to the public site link is that you can search the course from any page: assignments, submissions, progress report, anything.

ptn committed:
Added highlighted snippets

ptn commented:
@jordanbyron I ended up using a helper that wraps `highlight` instead of a complete different object with TextSearchResult. Does that look good?

I'm still working on putting a search bar in the nav, so you can search a course from any page, and I have a question about that. Since the form can't be remote, because the user might not be looking at the courses/show review, would it be OK if I made CoursesController#show also handle search parameters and leave #search to handle only ajax requests? This way, the nav search form would simply point to the show action.

Any suggestions? Or maybe nav search is a bad idea altogether?

jordanbyron commented:
> would it be OK if I made CoursesController#show also handle search parameters and leave #search to handle only ajax requests?

You can have `CoursesController#search` respond to both javascript and html requests. At the bottom of that method, just do something like this:

``` ruby
def search
  # Do some search related stuff
  respond_to do |format|
    format.html # Will by default render app/views/courses/search.html
    format.js   # Will by default render app/views/courses/search.js
  end
end
```

Does that help?

ptn commented:
Right now, search results are shown in a section of the courses/show view. If I used format.html from the search action, I would either need to write from scratch the `app/views/courses/search.html` view or repeat the logic of the show action so I can do `format.html { render :show }`

I'm not sure which one would be best. If we do `format.html { render :show }`, I could extract everything in the show action into a helper method and then invoke it from both show and search so that the template can render, but this implies that they are mostly the same thing, which is not good. This is why I thought of using search only for ajax, but that's not much better. Or maybe the search action could make the links to the right send a request for `show` instead of being anchor links, but this feels a lot like bolting the feature on to the side instead of nicely inserting it into the design.

If you were writing a search bar that's usable not only from courses/show, but also from assignments and submissions, how would you do it? Or better yet, what do you feel this patch still needs before merging?

ptn committed:
Redesign of the search results

ptn committed:
Fix form spanning the whole page width

ptn commented:
Hi @jordanbyron,

I just gave the UI a fresh coat of love. If I'm not mistaken, this finishes the original patch. What would be the next step?

Thanks!

ptn committed:
Added tests for the helper #highlighted_snippet

ptn committed:
Added tests for unscoped texticle searches

ptn commented:
Hi @jordanbyron,

I just pushed unit tests for the code I've written, how do they look?

Should we still go forward with the search bar in the nav? What do you think this patch still needs before merging?

jordanbyron commented:
Hey Pablo,

I've got to catchup on some of my other work before I can review this. If you really need someone to look at it you can ask @sandal or @semmons99 to review your code.

jordanbyron commented on app/models/course.rb:
```diff
@@ -92,6 +96,19 @@ def full_name
       sort_by { |a| a.created_at }.reverse
+  end                                     
+
+  def search_course_resources(search_key)
```
Let's rename this method to just `search`.

ptn replied:
Right away! I don't like it either, I kept the name I found. Will do!

jordanbyron commented on app/models/course.rb:
```diff
@@ -92,6 +96,19 @@ def full_name
+    results[:irc_messages] = search_irc_messages(search_key)
+
+    results
   end
```
Can you align all the `=` signs in this method? This is a very small thing, but it really helps with code readability.

jordanbyron commented on app/models/course.rb:
```diff
@@ -99,6 +116,41 @@ def activities
+  def search_assignments(search_key)
+    assignments.empty? ? [] :
+      Assignment.search(search_key, self.assignments)
+  end
```
I usually try to avoid ternary operators just because they make the code just a little harder to trace. In this case it might be clearer to do something like this:

``` ruby
def search_assignments(search_key)
  return [] if assignments.empty?
  Assignment.search(search_key, assignments)
end
```

That way it is very clear that an empty array is returned if there are no assignments. Same applies to `Course#search_submissions`

jordanbyron commented on app/views/courses/_search.html.haml:
```diff
@@ -0,0 +1,7 @@
+  :class => "bp search" do
+  = text_field_tag :search, params[:search], :placeholder => "Search current course"
+  = submit_tag 'Search', :class => "inline"
+.clear
```
We need to indicate that the search is running in the background. Spinner images work great for that. Surprisingly we don't have any in the project, but you can generate one over at [spin.js](http://fgnass.github.com/spin.js/) and add it to the project.

Then with a little javascript you can listen to the form's `ajax:before` and `ajax:complete` events to show and hide the image. If you need help with that just let me know.

jordanbyron commented on test/integration/courses/search_test.rb:
```diff
@@ -0,0 +1,66 @@
+      end      
+    end
+  end
+end
```
Is there a reason all these tests are pending? I'm guessing it's because we don't have an existing way to test javascript dependent actions in the project. I've been using `capybara-webkit` with great success. Feel free to add that to the Gemfile and see if you can get these tests working.

jordanbyron commented on app/views/courses/_search_results.html.haml:
```diff
@@ -0,0 +1,59 @@
+      - @results[:submissions].each do |submission|
+        %ul
+          %li
+            - link_text = submission.user.name + "'s submission for " + submission.assignment.name
```
Try not to define variables in the view templates if you can. This is probably a good case for a helper method.

jordanbyron commented on app/views/courses/_search_results.html.haml:
```diff
@@ -0,0 +1,59 @@
+      %hr
+      - @results[:submission_comments].each do |comment|
+        - submission = comment.commentable
+        - link_text = 'A comment in ' + submission.user.name + "'s submission for " + submission.assignment.name
```
Also another good candidiate for a helper method

jordanbyron commented:
Hi Pablo,

I think we are really close to being done with this feature. Take a look at my feedback and let me know if anything is unclear. Again sorry it took so long to get back to you. It's been a crazy week so far and I am still catching up :sweat:

Thanks for working on this :smile:

ptn committed:
Rename search_course_resources to simply search

ptn committed:
Align assignment operators

ptn committed:
Don't use multiline ternary operator

ptn committed:
Move logic from _search_results.html.haml to a helper

ptn committed:
Show JS spinner when searching a course

ptn committed:
Don't use br for spacing, do that with css

ptn committed:
The context class is not necessary in #search_results

ptn commented:
Hi Jordan,

I've corrected almost everything you pointed out, we are only missing the integration tests, but honestly, I'm a total noob with  everything test. Here's what I've done so far:
1. I added capybara-webkit to the Gemfile and ran `bundle`
2. I added this to `test_helper.rb` inside of `ActionDispatch::IntegrationTest`:
   
   ``` ruby
   Capybara.javascript_driver = :webkit
   
    teardown do
      Capybara.reset_sessions!
      Capybara.use_default_driver
    end
   ```
3. In `test/integration/courses/search_test.rb`, I added this line to `setup`:
   
   ``` ruby
   Capybara.current_driver = Capybara.javascript_driver
   ```
4. In the test itself, I'm doing this:
   
   ``` ruby
   visit course_path(@course, :anchor => "search")
   ```
   
   where `@course` is the result of `@course = Factory(:course)`

And I get the following error:

```
Capybara::Driver::Webkit::WebkitError: Unable to load URL: http://127.0.0.1:49716/courses/73#search
```

Any hints? Am I setting capybara-webkit wrong?

Thanks!

jordanbyron commented:
Pablo,

Push your integration test code so I can look at it and see if it runs on my machine.

ptn committed:
Integration setup and test run

ptn commented:
OK, pushed. I didn't want to push code that doesn't run, but I guess it's easier to compare this way.

Does it run on your machine?

jordanbyron commented:
> I didn't want to push code that doesn't run, but I guess it's easier to compare this way.

That's ok. We're using this pull request to discuss your code and it's hard to do that when you don't push the code we're talking about :wink:

> Does it run on your machine?

Nope, but I was able to figure out why. Usually when tests are failing like this it usually means you're not testing the page you think you are. I was able to confirm that suspicion by adding one line to the test:

``` ruby
test "I don't get result if there is no match" do
  visit course_path(@course)

  raise current_path.inspect

  fill_in 'search', with: 'blablabla'
  click_button('Search')
  assert_content 'Found 0 results'
end
```

When I ran the tests, this is what I got:

```
RuntimeError: "/users/sign_in"
```

So as you can see, the user isn't being logged in. Which is strange because the other integration tests are working. So next thing I did was to install `capybara-screenshot` so I can see what's going on. With that working I could see that the username and password combination were being rejected. Again that is odd because I'd expect that to work as it did with all our other integration tests. I even double checked to ensure the `User` record was really saved to the database.

Then I remembered reading something in `capybara-webkit`'s README:

> Take note of the transactional fixtures section of the [capybara README](https://github.com/jnicklas/capybara/blob/master/README.md)

So I took a look there and found this:

> Some Capybara drivers need to run against an actual HTTP server. Capybara takes care of this and starts one for you in the same process as your test, but on another thread. Selenium is one of those drivers, whereas RackTest is not.
> 
> If you are using an SQL database, it is common to run every test in a transaction, which is rolled back at the end of the test, rspec-rails does this by default out of the box for example. Since transactions are usually not shared across threads, this will cause data you have put into the database in your test code to be invisible to Capybara.
> 
> Cucumber handles this by using truncation instead of transactions, i.e. they empty out the entire database after each test. You can get the same behaviour by using a gem such as database_cleaner.

So with `capybara-webkit` that's what we need to do. And since this is starting to go way beyond the scope of this pull request, I'll work on bringing our test harness up-to-date and then you should be able to merge in my changes, and your tests will most likely work. I'll let you know once that is done. :sweat:

jordanbyron commented:
Ok @ptn you should be able to merge in my changes from master, fix any conflicts, and get your tests working. Let me know if you have any problems with the merge. You can always double check what is in master now by browsing the code on github.

ptn committed:
Merge branch 'master' into search

Conflicts:
	Gemfile
	Gemfile.lock
	test/test_helper.rb

ptn commented:
> So as you can see, the user isn't being logged in. Which is strange because the other integration tests are working. So

I suspected the same, but then I grepped sign_user_in and saw it all over the place, so I figured it had to be something else. I read about database cleaner, but since I always see CREATE TABLE statements when I run the tests, I thought that you were doing that some other way.

Anyway, it's  working now! I'll work on the other tests and push them up as soon as I can.

BTW, to pull in your changes I merged them into this branch, because I didn't want to rebase what I've already pushed. We now have a merge commit, is this OK or should I rebase anyway?

jordanbyron commented:
> We now have a merge commit, is this OK or should I rebase anyway?

That's ok.

ptn committed:
Second integration test for Course#search

Uses new test helpers and updates the setup method

ptn committed:
Finish integration tests for Course#search

ptn committed:
Fix showing matched irc messages when invoking Course#search

ptn committed:
Reuse search_results partial from chat messages search in courses search

ptn commented:
@jordanbyron tests complete!

Please take a look at that last commit, I had to shuffle some things around to reuse a partial for the chat messages search in the course search. Does that look OK? If so, then I think we are done, right?

Commit:
Update Gemfile

Commit:
Fix typo in tests

Commit:
Minor modifications to search view

Commit:
Because we rip people from courses, we have orphaned comments

Commit:
Highlight results should be nice and bright

Commit:
Link to the actual comment

Commit:
Small code cleanup in Course model

Commit:
Add failing tests for Course#search #134

jordanbyron commented:
Pablo:

Unfortunately I think there is a problem with the scoped search. Check out the [failing tests](02d6eac26ff90345f396dc58ee20ee0c8b4c280d) I just pushed to the [ptn-search branch](https://github.com/mendicant-university/university-web/tree/ptn-search). I can't find anything in texticle's docs that explains how we're using the search method:

``` ruby
Assignment::Submission.search(search_key, submissions)
```

How did you find that? Or was that something @Nerian worked on?

ptn committed:
Assign course to channel in test setup

ptn committed:
Correctly scope texticle searches using a where clause

ptn commented:
Jordan,

Yeah, that's something of @Nerian's. I mostly saw it working so I never checked or dug deeper, my bad, but now I see that what texticle actually expects as a second parameter is a boolean. I previously thought that what it expected was a subset of what records to search and if it was an empty collection it simply ignored it, so my first fix was not a fix at all.

Since texticle returns an ActiveRecord Relation, I am now scoping simply by chaining into a where clause. All tests pass now.

Let me know what you think. Thanks!

jordanbyron commented:
And with that fixed I think this feature is ready to be merged! :balloon: :tada: :balloon:

Thanks for doing such a great job finishing this feature up. I'll merge it into master as soon as the January core course wraps up. :smile:

ptn commented:
\o/ !!!

Thanks Jordan, this has been great. You are an awesome maintainer, I really enjoyed working on this!

I plan to continue contributing for a while after the course is done and also I'd like to work on Mission of Mercy, I'm looking forward to that!
